[DEVEL_X] Implicitly-declared 'edm::View<int>::View(const edm::View<int>&)' is deprecated

Build log: [link](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_14_0_DEVEL_X_2023-11-15-2300/DataFormats/Common).

Copying the error message (as the log will disappear at some point)
```
>> Compiling  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/testRunner.cpp
CMSSW_CPU_TYPE= /data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_0_DEVEL_X_2023-11-15-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_0_DEVEL_X_2023-11-15-2300' -I/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/boost/1.80.0-826a207b8543c52970cb1f72d50f068c/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/cppunit/1.15.x-fb84a4bbf5a436317d208e3ef0864e91/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/lcg/root/6.26.11-1ef575234f3b954353280f3578660947/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tbb/v2021.9.0-e755918dac6a30ec36eff63ac4f7ddec/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/fmt/8.0.1-54e94b39f5cf29341bb9c4765764e1ca/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tinyxml2/6.2.0-d17873b4d6a42a43226cf689f82ec1ef/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr  -fPIC  -MMD -MF tmp/el8_amd64_gcc12/src/DataFormats/Common/test/testDataFormatsCommon/testRunner.cpp.d /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/testRunner.cpp -o tmp/el8_amd64_gcc12/src/DataFormats/Common/test/testDataFormatsCommon/testRunner.cpp.o
In file included from /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h:18,
                 from /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/ref_t.cppunit.cc:16:
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h: In instantiation of 'std::unique_ptr<edm::ViewBase> edm::View<T>::doClone() const [with T = {anonymous}::IntValue]':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:293:29:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:294:38: warning: implicitly-declared 'edm::View<{anonymous}::IntValue>::View(const edm::View<{anonymous}::IntValue>&)' is deprecated [-Wdeprecated-copy]
   294 |     return std::unique_ptr<ViewBase>{new View(*this)};
      |                                      ^~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:298:19: note: because 'edm::View<{anonymous}::IntValue>' has user-provided 'edm::View<T>& edm::View<T>::operator=(const edm::View<T>&) [with T = {anonymous}::IntValue]'
  298 |   inline View<T>& View<T>::operator=(View<T> const& rhs) {
      |                   ^~~~~~~
In file included from /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/reftobaseprod_t.cppunit.cc:14:
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h: In instantiation of 'edm::RefToBaseProd<T>::RefToBaseProd(const edm::RefToBaseProd<T>&) [with T = {anonymous}::Dummy]':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/reftobaseprod_t.cppunit.cc:169:44:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h:111:47: warning: implicitly-declared 'edm::View<{anonymous}::Dummy>::View(const edm::View<{anonymous}::Dummy>&)' is deprecated [-Wdeprecated-copy]
   111 |       product_.setProductPtr(ref.viewPtr() ? (new View<T>(*ref)) : nullptr);
      |                                              ~^~~~~~~~~~~~~~~~~~
In file included from /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h:18:
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:298:19: note: because 'edm::View<{anonymous}::Dummy>' has user-provided 'edm::View<T>& edm::View<T>::operator=(const edm::View<T>&) [with T = {anonymous}::Dummy]'
  298 |   inline View<T>& View<T>::operator=(View<T> const& rhs) {
      |                   ^~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h: In instantiation of 'edm::RefToBaseProd<T>::RefToBaseProd(const edm::RefToBaseProd<T>&) [with T = test_with_dictionaries::IntValue]':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/test/reftobaseprod_t.cppunit.cc:231:42:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/RefToBaseProd.h:111:47: warning: implicitly-declared 'edm::View<test_with_dictionaries::IntValue>::View(const edm::View<test_with_dictionaries::IntValue>&)' is deprecated [-Wdeprecated-copy]
   111 |       product_.setProductPtr(ref.viewPtr() ? (new View<T>(*ref)) : nullptr);
      |                                              ~^~~~~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:298:19: note: because 'edm::View<test_with_dictionaries::IntValue>' has user-provided 'edm::View<T>& edm::View<T>::operator=(const edm::View<T>&) [with T = test_with_dictionaries::IntValue]'
  298 |   inline View<T>& View<T>::operator=(View<T> const& rhs) {
      |                   ^~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h: In instantiation of 'std::unique_ptr<edm::ViewBase> edm::View<T>::doClone() const [with T = {anonymous}::Dummy]':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:293:29:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:294:38: warning: implicitly-declared 'edm::View<{anonymous}::Dummy>::View(const edm::View<{anonymous}::Dummy>&)' is deprecated [-Wdeprecated-copy]
   294 |     return std::unique_ptr<ViewBase>{new View(*this)};
      |                                      ^~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:298:19: note: because 'edm::View<{anonymous}::Dummy>' has user-provided 'edm::View<T>& edm::View<T>::operator=(const edm::View<T>&) [with T = {anonymous}::Dummy]'
  298 |   inline View<T>& View<T>::operator=(View<T> const& rhs) {
      |                   ^~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h: In instantiation of 'std::unique_ptr<edm::ViewBase> edm::View<T>::doClone() const [with T = test_with_dictionaries::IntValue]':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:293:29:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:294:38: warning: implicitly-declared 'edm::View<test_with_dictionaries::IntValue>::View(const edm::View<test_with_dictionaries::IntValue>&)' is deprecated [-Wdeprecated-copy]
   294 |     return std::unique_ptr<ViewBase>{new View(*this)};
      |                                      ^~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/225a463d1231a8a4c99dd9e7e0b20bde/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_DEVEL_X_2023-11-15-2300/src/DataFormats/Common/interface/View.h:298:19: note: because 'edm::View<test_with_dictionaries::IntValue>' has user-provided 'edm::View<T>& edm::View<T>::operator=(const edm::View<T>&) [with T = test_with_dictionaries::IntValue]'
  298 |   inline View<T>& View<T>::operator=(View<T> const& rhs) {
      |                   ^~~~~~~
```

--- a/DataFormats/Common/interface/View.h
+++ b/DataFormats/Common/interface/View.h
@@ -46,9 +46,11 @@ namespace edm {
     std::unique_ptr<ViewBase> clone() const;
 
   protected:
-    ViewBase();
-    ViewBase(ViewBase const&);
-    ViewBase& operator=(ViewBase const&);
+    ViewBase() = default;
+    ViewBase(ViewBase const&) = default;
+    ViewBase(ViewBase&&) = default;
+    ViewBase& operator=(ViewBase const&) = default;
+    ViewBase& operator=(ViewBase&&) = default;
     virtual std::unique_ptr<ViewBase> doClone() const = 0;
     void swap(ViewBase&) {}  // Nothing to swap
   };
@@ -101,11 +103,7 @@ namespace edm {
     // infrastructure code.
     View(std::vector<void const*> const& pointers, FillViewHelperVector const& helpers, EDProductGetter const* getter);
 
-    ~View() override;
-
     void swap(View& other);
-
-    View& operator=(View const& rhs);
 
     size_type capacity() const;
 
@@ -190,9 +188,6 @@ namespace edm {
       }
     }
   }
-
-  template <typename T>
-  View<T>::~View() {}
 
   template <typename T>
   inline void View<T>::swap(View& other) {
@@ -292,13 +287,6 @@ namespace edm {
   template <typename T>
   std::unique_ptr<ViewBase> View<T>::doClone() const {
     return std::unique_ptr<ViewBase>{new View(*this)};
-  }
-
-  template <typename T>
-  inline View<T>& View<T>::operator=(View<T> const& rhs) {
-    View<T> temp(rhs);
-    this->swap(temp);
-    return *this;
   }
 
   template <typename T>

--- a/DataFormats/Common/src/View.cc
+++ b/DataFormats/Common/src/View.cc
@@ -18,8 +18,4 @@ namespace edm {
     return p;
   }
 
-  ViewBase::ViewBase() {}
-
-  ViewBase::ViewBase(ViewBase const&) {}
-
 }  // namespace edm

--- a/DataFormats/Common/test/ptr_t.cppunit.cc
+++ b/DataFormats/Common/test/ptr_t.cppunit.cc
@@ -206,6 +206,9 @@ void testPtr::constructTest() {
     Ptr<Dummy> copyPtr(dummy2Ptr);
     CPPUNIT_ASSERT(dummy2Ptr.key() == copyPtr.key());
     CPPUNIT_ASSERT(dummy2Ptr.get() == static_cast<const Dummy*>(dummy2Ptr.get()));
+
+    Ptr<Dummy> movePtr(std::move(copyPtr));
+    CPPUNIT_ASSERT(dummy2Ptr.key() == movePtr.key());
   }
 }
 

--- a/DataFormats/Common/test/reftobaseprod_t.cppunit.cc
+++ b/DataFormats/Common/test/reftobaseprod_t.cppunit.cc
@@ -173,6 +173,10 @@ void testRefToBaseProd::constructTest() {
 
     CPPUNIT_ASSERT(dummyPtr2.id() == pid);
     compareTo(dummyPtr2, dummyContainer);
+
+    RefToBaseProd<Dummy> dummyPtr3(std::move(dummyPtr2));
+    CPPUNIT_ASSERT(dummyPtr3.id() == pid);
+    compareTo(dummyPtr3, dummyContainer);
   }
 }
 

